Moved Analytics to a different module

jawnpaul opened the pull request:
Integrate Mixpanel Analytics and Refactor Analytics Logging

- Added Mixpanel integration for analytics alongside Firebase.
- Created a new `analytics` module to encapsulate analytics logging.
- Introduced a `CompositeLogger` to route analytics events to multiple providers (Firebase and Mixpanel).
- Updated the app to use the `CompositeLogger` for logging events such as default prompt selection, company selection, and content sharing/saving.
- Removed the previous analytics logging implementation from the app module.
- Added ProGuard rules for the `analytics` module to ensure proper code optimization and prevent obfuscation of necessary classes and methods.
- Updated local properties and build configurations to include Mixpanel API tokens for development and production environments.

jawnpaul committed:
feat: Integrate Mixpanel Analytics and Refactor Analytics Logging

- Added Mixpanel integration for analytics alongside Firebase.
- Created a new `analytics` module to encapsulate analytics logging.
- Introduced a `CompositeLogger` to route analytics events to multiple providers (Firebase and Mixpanel).
- Updated the app to use the `CompositeLogger` for logging events such as default prompt selection, company selection, and content sharing/saving.
- Removed the previous analytics logging implementation from the app module.
- Added ProGuard rules for the `analytics` module to ensure proper code optimization and prevent obfuscation of necessary classes and methods.
- Updated local properties and build configurations to include Mixpanel API tokens for development and production environments.

jawnpaul committed:
feat: Integrate Mixpanel Analytics and Refactor Analytics Logging

- Added Mixpanel integration for analytics alongside Firebase.
- Created a new `analytics` module to encapsulate analytics logging.
- Introduced a `CompositeLogger` to route analytics events to multiple providers (Firebase and Mixpanel).
- Updated the app to use the `CompositeLogger` for logging events such as default prompt selection, company selection, and content sharing/saving.
- Removed the previous analytics logging implementation from the app module.
- Added ProGuard rules for the `analytics` module to ensure proper code optimization and prevent obfuscation of necessary classes and methods.
- Updated local properties and build configurations to include Mixpanel API tokens for development and production environments.

jawnpaul committed:
Merge remote-tracking branch 'origin/analytics' into analytics

# Conflicts:
#	analytics/src/main/java/com/thejawnpaul/gptinvestor/analytics/di/AnalyticsModule.kt
#	analytics/src/main/java/com/thejawnpaul/gptinvestor/analytics/mixpanel/MixpanelLogger.kt

jawnpaul committed:
Fix: Use correct Mixpanel tokens for debug and release builds

-   Swapped the Mixpanel tokens used for debug and release builds.
-   The debug build now uses the development token ("MIXPANEL_DEV_TOKEN"), and the release build uses the production token ("MIXPANEL_PROD_TOKEN").

jawnpaul committed:
feat: Add top pick selected analytics event

- Adds a `logTopPickSelected` event to the `Analytics` interface and its implementations (`FirebaseLogger`, `MixpanelLogger`, `CompositeLogger`).
- Logs the top pick selection event with company ticker and name in `TopPickRepository` when fetching a top pick.

jawnpaul committed:
Refactor: Replace CompositeLogger with Analytics interface

- Replaced direct usage of `CompositeLogger` with the `Analytics` interface in `TopPickRepository`, `ConversationRepository`, and `CompanyRepository`.
- Modified `provideDefaultAnalyticsLogger` in `AnalyticsBindingModule` to return an `Analytics` instance instead of `CompositeLogger`.
- Added `@Inject` annotation to the `CompositeLogger` class constructor.

jawnpaul committed:
feat: Refactor analytics events and parameters

- Updated `Analytics` interface to use a single `logEvent` method with event name and parameters, and a `logViewEvent` for screen views.
- Updated `MixpanelLogger`, `FirebaseLogger`, and `CompositeLogger` to reflect the new `logEvent` and `logViewEvent` methods.
- Updated `FirebaseLogger` to use `putString` for all bundle parameters to ensure compatibility.
- Modified `MixpanelLogger` to handle reserved event names ("sign up", "log in", "log out", "delete account") separately.
- Updated repositories (`TopPickRepository`, `ConversationRepository`, `CompanyRepository`) to use the new `logEvent` method with appropriate event names and parameters.

Ezike commented on app/src/main/java/com/thejawnpaul/gptinvestor/features/company/data/repository/CompanyRepository.kt:
```diff
@@ -24,7 +24,7 @@ import timber.log.Timber
     private val apiService: ApiService,
     private val companyDao: CompanyDao,
-    private val analyticsLogger: AnalyticsLogger
+    private val analyticsLogger: Analytics
```
```suggestion
    private val analyticsLogger: AnalyticsLogger
```

Ezike commented on analytics/src/main/java/com/thejawnpaul/gptinvestor/analytics/mixpanel/MixpanelLogger.kt:
```diff
@@ -0,0 +1,47 @@
+        params: Map<String, Any>
+    ) {
+        when (eventName.lowercase()) {
+            "sign up" -> {
```
What are the sign up, etc for

jawnpaul replied:
They are special type of events that require extra logic. For example, in the case of mixpanel we don't call `track` to log a sign up event but rather we use `identify` as explained [here](https://docs.mixpanel.com/docs/tracking-methods/sdks/android#identify). Also, for log out, we call `reset` as explained [here](https://docs.mixpanel.com/docs/tracking-methods/sdks/android#call-reset-at-logout)

Ezike replied:
That’s interesting. Thanks. 
i wonder if the logger should expose the reset and identify apis instead of checking for event types here.

Ezike replied:
It was a suggestion sha. It’s not clear that it’s the better approach.

jawnpaul replied:
😂 I've already implemented it. The suggestion sounded nice

Ezike replied:
It’s a fair enough idea but It also bloats the api exposing more functionality than most screens require. Managing user sessions could be a separate entity altogether. one other idea is also this 
```kotlin
interface AuthLogger:AnalyticsLogger {
      fun identify(…)
      fun reset(…)
}
```

Ezike commented on analytics/src/main/java/com/thejawnpaul/gptinvestor/analytics/composite/CompositeLogger.kt:
```diff
@@ -0,0 +1,68 @@
+        }
+    }
+
+    class Builder @Inject constructor(
```
Why is the builder still needed

Ezike commented on analytics/src/main/java/com/thejawnpaul/gptinvestor/analytics/Analytics.kt:
```diff
@@ -0,0 +1,12 @@
+    fun logEvent(eventName: String, params: Map<String, Any>)
+    fun logViewEvent(screenName: String)
+
+    enum class Provider {
```
Provider should probably not be declared in the public api

Ezike commented on analytics/src/main/java/com/thejawnpaul/gptinvestor/analytics/composite/CompositeLogger.kt:
```diff
@@ -0,0 +1,68 @@
+import javax.inject.Singleton
+
+@Singleton
+class CompositeLogger @Inject constructor(private val loggers: Map<Analytics.Provider, Analytics>) :
```
Why not inject firebase n mix panel instances here instead of wrapping them in a class

jawnpaul committed:
feat: Refactor Analytics Interface and Implementations

- Renamed `Analytics` interface to `AnalyticsLogger` for clarity.
- Updated `MixpanelLogger`, `FirebaseLogger`, and `CompositeLogger` to implement `AnalyticsLogger`.
- Modified repositories (`TopPickRepository`, `ConversationRepository`, `CompanyRepository`) to use `AnalyticsLogger`.
- Simplified `CompositeLogger` to directly use `FirebaseLogger` and `MixpanelLogger` instances.
- Updated dependency injection in `AnalyticsBindingModule` to bind `CompositeLogger` to `AnalyticsLogger`.

jawnpaul committed:
feat: Implement User Identification and Reset in Analytics Loggers

- Implemented `identifyUser` and `resetUser` functions in `MixpanelLogger` to track and reset user identification in Mixpanel.
- Implemented `identifyUser` and `resetUser` functions in `FirebaseLogger` to set user properties and reset analytics data in Firebase.
- Added `identifyUser` and `resetUser` functions to the `AnalyticsLogger` interface.

jawnpaul committed:
feat: Implement User Identification and Reset in CompositeLogger

- Implemented `identifyUser` and `resetUser` methods in `CompositeLogger` to delegate user identification and reset actions to both Firebase and Mixpanel loggers.

Ezike commented on analytics/src/main/java/com/thejawnpaul/gptinvestor/analytics/di/AnalyticsBindingModule.kt:
```diff
@@ -0,0 +1,41 @@
+
+@Module
+@InstallIn(SingletonComponent::class)
+abstract class AnalyticsBindingModule {
```
This can be an interface

Ezike commented on analytics/src/main/java/com/thejawnpaul/gptinvestor/analytics/mixpanel/MixpanelLogger.kt:
```diff
@@ -0,0 +1,44 @@
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
```
Why singleton
